File: me/src/models/order.rs
use redis::RedisResult;
use std::collections::HashMap;

#[derive(Clone, Hash, Eq, PartialEq, Debug)]
pub struct Order {
    pub id: String,
    pub product_id: String,
    pub user_id: String,
    pub side: String,
    pub r#type: String,
    pub created_at: i64,
    pub executed_value: i64,
    pub status: String,
    pub settled: bool,
    pub price: Option<i64>,
    pub cancel_after: Option<String>,
    pub size: i64,
}

impl Order {
    pub fn from_redis_map(
        map: &HashMap<String, redis::Value>,
    ) -> RedisResult<Self> {
        Ok(Self {
            id: redis::from_redis_value(map.get("id").unwrap())?,
            product_id: redis::from_redis_value(
                map.get("product_id").unwrap(),
            )?,
            user_id: redis::from_redis_value(map.get("user_id").unwrap())?,
            side: redis::from_redis_value(map.get("side").unwrap())?,
            r#type: redis::from_redis_value(map.get("type").unwrap())?,
            created_at: redis::from_redis_value(
                map.get("created_at").unwrap(),
            )?,
            executed_value: redis::from_redis_value(
                map.get("executed_value").unwrap(),
            )?,
            status: redis::from_redis_value(map.get("status").unwrap())?,
            settled: {
                redis::from_redis_value::<String>(map.get("settled").unwrap())
                    .unwrap()
                    .parse::<bool>()
                    .unwrap()
            },
            price: {
                map.get("price")
                    .and_then(|v| redis::from_redis_value::<i64>(v).ok())
            },
            cancel_after: {
                map.get("cancel_after")
                    .and_then(|v| redis::from_redis_value::<String>(v).ok())
            },
            size: redis::from_redis_value(map.get("size").unwrap())?,
        })
    }

    pub fn to_redis_tuples(&self) -> Vec<(&str, String)> {
        let mut fields = vec![
            ("id", self.id.clone()),
            ("product_id", self.product_id.clone()),
            ("user_id", self.user_id.clone()),
            ("side", self.side.clone()),
            ("type", self.r#type.clone()),
            ("created_at", self.created_at.to_string()),
            ("status", self.status.clone()),
            ("executed_value", self.executed_value.to_string()),
            ("settled", self.settled.to_string()),
            ("size", self.size.to_string()),
        ];

        // Option fields
        if let Some(price) = self.price {
            fields.push(("price", price.to_string()));
        }

        if let Some(cancel_after) = &self.cancel_after {
            fields.push(("cancel_after", cancel_after.clone()));
        }

        fields
    }
}

File: md/src/models/order.ts
export class Order {
  public id: string;
  public product_id: string;
  public user_id: string;
  public side: 'buy' | 'sell';
  public type: 'market' | 'limit';
  public created_at: number;
  public executed_value: number;
  public status: 'received' | 'open' | 'done';
  public settled: boolean;
  public price: number;
  public cancel_after: 'min' | 'hour';
  public size: number;

  constructor(data: Record<string, string>) {
    this.id = data.id;
    this.product_id = data.product_id;
    this.user_id = data.user_id;
    this.side = data.side as 'buy' | 'sell';
    this.type = data.type as 'market' | 'limit';
    this.created_at = parseInt(data.created_at, 10);
    this.executed_value = parseFloat(data.executed_value);
    this.status = data.status as 'received' | 'open' | 'done';
    this.settled = data.settled === 'true';
    this.price = parseInt(data.price, 10);
    this.cancel_after = data.cancel_after as 'min' | 'hour';
    this.size = parseInt(data.size, 10);
  }
}
